content/en/blog/_posts/2021-09-06-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,85 @@
+        <br>
+        <i>We map it to <code>requests.memory</code></i>
+        </td>
+   </tr>
+   <tr>
+       <td>memory.high</td>
+       <td><code>memory.high</code> is the memory usage throttle limit. This is the main mechanism to control a cgroup's memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. 
+       <br>
+       <br>
+       <i>We use a formula to calculate <code>memory.high</code>, depending on <code>limits.memory</code> or <code>node allocatable memory</code> and a  throttling factor</i>
+       </td>
+   </tr>
+</table>
+
+When container memory requests are made, kubelet passes `memory.min` to the back-end CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during container creation. 
+
+![](./container-memory-min.svg)
+
+Since the `memory.min` interface requires that the ancestor cgroup directories are all set, the pod and node cgroup directories need to be set rightly. 
+
+![](./pod-memory-min.svg)
+![](./node-memory-min.svg)
+
+However, unlike container, kubelet communicates with cgroup v2 directly through libcontainer of runc for pod and node level.
+
+For memory limits, in addition to the original way of limiting memory usage, Memory QoS adds an additional feature of throttling memory allocation. A throttling factor is introduced as a multiplier (default is 0.8). If the result of multiplying memory limits by the factor is greater than memory requests, kubelet will set `memory.high` to the value and use `Unified` via CRI. And if the container does not specify memory limits, kubelet will use node allocatable memory instead. 
+
+![](./container-memory-high.svg)
+
+## How do I use it?
+Here are the prerequisites for enabling Memory QoS on your Linux node, some of these are related to [Kubernetes support for cgroup v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2).
+
+1. Kubernetes since v1.22
+2. [runc](https://github.com/opencontainers/runc) since v1.0.0-rc93; [containerd](https://containerd.io/) since 1.14
+3. Linux kernel minimum version: 4.15, recommended version: 5.2+
+4. You have enabled the cgroup v2 unified hierarchy (for example: add `systemd.unified_cgroup_hierarchy=1` to the kernel command line)
+
+Container OCI has already supported cgroup v2 [`Unified`](https://github.com/opencontainers/runtime-spec/blob/master/config-linux.md#unified), and Kubernetes CRI has also made the desired changes to support passing [`Unified`](https://github.com/kubernetes/kubernetes/pull/102578). However, CRI Runtime support is needed here else. Memory QoS in Alpha phase is designed to support containerd and cri-o. [Changes](https://github.com/containerd/containerd/pull/5627) to containerd have been committed and ready for review. Users can temporarily merge MR and compile the version before the code merge. cri-o is still in WIP.
```

Comment:
Container OCI -> OCI runtimes such as runc and crun already support cgroups v2.

Reply:
Link to CRI-O pull request is https://github.com/cri-o/cri-o/pull/5207

Reply:
done.